frontend/src/metabase/writeback/components/ActionCreator/FormCreator/FormCreator.tsx

```diff
@@ -1,38 +1,92 @@
-import React from "react";
+import React, { useState, useEffect } from "react";
 import { t } from "ttag";
 
-import Icon from "metabase/components/Icon";
-
 import type { TemplateTag } from "metabase-types/types/Query";
+import type {
+  ActionFormSettings,
+  FieldSettings,
+} from "metabase/writeback/types";
+
+import { FieldSettingsPopover } from "./FieldSettingsPopover";
+import { getDefaultFormSettings, getDefaultFieldSettings } from "./utils";
+import { FormField } from "./FormField";
 
 import {
   FormItemWrapper,
   FormCreatorWrapper,
   FormItemName,
-  Input,
   FormSettings,
   EmptyFormPlaceholderWrapper,
 } from "./FormCreator.styled";
 
-export function FormCreator({ tags }: { tags: TemplateTag[] }) {
+export function FormCreator({
+  tags,
+  formSettings: passedFormSettings,
+  onChange,
+}: {
+  tags: TemplateTag[];
+  formSettings?: ActionFormSettings;
+  onChange: (formSettings: ActionFormSettings) => void;
+}) {
+  const [formSettings, setFormSettings] = useState<ActionFormSettings>(
+    passedFormSettings?.fields ? passedFormSettings : getDefaultFormSettings(),
+  );
+
+  useEffect(() => {
+    onChange(formSettings);
```

Comment:
I wanted to keep the actual state at this level where the object is pretty simple, and just pass updates up one-way for performance reasons. updating the `visualization_settings` in the Question object itself could get really slow.

Reply:
Is it that slow? We do that all the time in the QB I think

Reply:
with large numbers of fields, that pattern in the filter modal incurs some UI lag.